### DISCUSSION TITLE
Most applications crash when using them

### DISCUSSION THREAD

**Discussion by Mikailo** [neurodesk/neurodesk.github.io#106] [Q&A] [answered]
Hello everyone. I am using the "20220329" version of neurodesk, and I have come across some issues whose cause I can't figure out. Since I haven't seen any prevalent issues anywhere, I assume it must be something with the environment, but I am not sure.

I encountered this issue when using.
    -Visual Studio Code: Even opened on its own after a while it crashes giving a code 11 or code 5 , for which I also cant find any info about them and what they mean.
    - MNE - 0.23.4 and 1.0.0: I followed the tutorial found here https://www.neurodesk.org/tutorials/electrophysiology/eeg_mne-python/ , and I can't finish it because in random intervals Visual Studio Code crashes, sometimes before even running anything others after running.
    - Freeview 7.2.0  crashes whenever I try loading anything 
    - Freeview 7.1.1: Trying to load results or recon all from freesurfer. Can  successfully load volumes, and some surfaces but other surfaces crash, while some surfaces may or may not crash randomly.
    - Freesurfer: haven't encountered any crashes yet.

This behaviour seems pretty odd and i cannot pinpoint the issue, I tried this in 3 different windows pc's ram, running from windows in a docker container and this behaviour seems the same in all of them.   
Do you have any clue why this is happening and potential fixes?

**Comment by stebo85** [comment]
Dear @Mikailo,

Very interesting to hear! There might be a few issues going on. Let's try to unpack this:

1) Random crashes point towards memory issues, so vscode randomly crashing could be a lack of shm memory. How are you starting neurodesk? What is the exact command you are using?

2) Freeview behaving differently in the different versions could point to a bug in freeview itself. Can you share an example file that crashes when you are opening it?

3) To exclude any setup issues on your side, can you try to run the problematic things here in our hosted version of neurodesk? https://play.neurodesk.org/v2/gh/neurodesk/jupyter-neurodesktop-image/main?urlpath=neurodesktop


Thank you so much for finding these bugs :)
 
Cheers
Steffen

**Reply by Mikailo** [reply]
Hello. first of all thanks for your assistance.

Regarding 2) and 3),  I tried replicating the behaviour in the hosted version and indeed could reproduce some issues. Specifically with freeview and indeed the exact same things happened. I have uploaded the files I used in my google drive here [https://drive.google.com/drive/folders/104qegfGeVvz3-66dqngmwWSmS_Nybsay?usp=sharing](url)  , I am not sure if this is the best ways to share them with you. Please do tell me if you have any other suggestions. 

Regarding 1) Indeed, this is what  I thought too, but I checked the available memory inside the container with watch, and outside through docker and my pc's and there seemed to be plenty, even run a memtest that used more memory, and there was no error as far as I could tell. To deploy it I use docker compose which follows the docker command found in getting started. I have also uploaded that in the previous link if you would to take a look. 

Should I try a previous version, I tried looking for the previous stable one, but wasn't certain which one to use, or should I try it without compose?

**Comment by stebo85** [comment, marked as answer]
Dear @Mikailo,

Thank you for providing the startup commands you are using and it confirms my suspicion - you are using an incorrect startup command that doesn't provide SHM memory to the docker container. Can you please adjust your docker compose command to reflect the startup command from our documentation?

```
docker run --shm-size=1gb -it --privileged --name neurodesktop -v C:/neurodesktop-storage:/neurodesktop-storage -p 8080:8080 -h neurodesktop-20220329 vnmd/neurodesktop:20220329
```

The --shm-size=1gb is crucial for everything around chrome/electron/vscode to work correctly - if you are not providing SHM in windows things will crash quickly as you experienced.


The issue with freeview is a bug in freeview itself - I will update the freesurfer container and hope it's fixed in the latest version.


All the best
Steffen

**Reply by Mikailo** [reply]
Thanks a lot @stebo85 , yeah I missed it, I wasn't aware of shared memory in docker. 
I had an idea that maybe freeview has an issue , since the freesurfer used was a different version. I will create new files using the latest one and see if this resolves the issue. Will inform you if that fixes it. Thanks again.

**Reply by stebo85** [reply]
Dear @Mikailo - I had a detailed look at freeview 7.2.0 and I can open your files without any issue. However, I am thinking that you might have missed that you NEED a license for Freeview to work and unfortunately, we are not allowed to distribute a license with Neurodesk!

So here is how you can run freeview 7.2.0 and open your files:

apply for a license (https://surfer.nmr.mgh.harvard.edu/registration.html) and paste this license in ~/.license

then run 
```
echo "export FS_LICENSE=~/.license" >> ~/.bashrc
```
then start freeview 7.2.0 and it should all work.



Can you let us know if this works for you?

**Reply by Mikailo** [reply]
Hello, yes indeed the issue was with the license, I mixed the ways to provide a license when using freesurfer on its own and I wasn't seeng the error. Thank you and sorry if I wasted your time, with things that I missed!.

**Reply by stebo85** [reply]
No worries @Mikailo :) we are aiming to make things as easy as possible and people missing things in the documentation is a sign that our documentation is not good enough. 

Based on your feedback I removed the Freeview link from the application menu so that people are forced to start Freeview from the terminal and with this see the license error clearly. Also, when starting the freesurfer container there is a detailed description about freesurfers license problems.

If you have any other ideas how to improve the documentation please submit suggestions and pull requests with clarifications - neurodesk is a community project :)

**Reply by Mikailo** [reply]
Yes thank you very much, indeed this way, I think I would have noticed the license issue easier.
The only other thing that caused my an issue, would be providing a ready docker compose file instead of only a docker command, but I am not sure if it is necessary and its just another thing to keep updated, so I am not sure if it is worth it.
Thank again.

**Reply by stebo85** [reply]
@Mikailo - why would you need a docker compose file? Happy to provide one if I can understand why someone would need this over the normal docker command :) can you post your working compose file to be included in the documentation?

Great to hear that the license error is now more obvious :)

Thank you
Steffen
